### ISSUE TITLE
Fix ParameterSet/VectorDistribution/ParameterSetCollection bottleneck

### ISSUE THREAD

**Issue by RaphaelS1** [alan-turing-institute/distr6#219]

**Comment by dcnorris** [comment]
Apropos of this issue, let me share how this performance issue has affected my own use of distr6.
This was my first time using R6 classes, and I think I may have treated them as more 'lightweight' than I should have.

The applicaton is in package [precautionary](https://CRAN.R-project.org/package=precautionary), where I have tried to use S4 classes to convey a relationship between a hyperprior (`mtdi_generator`) and the `mtdi_distributions` it gives rise to:

```r
setClass("mtdi_generator",
  slots = list(
    units = "character"
  ),
  contains = "VIRTUAL")

setClass("mtdi_distribution",
  slots = list(
    CV = "numeric"      # Establish CV and median as universal parameters
  , median = "numeric"  # for specifying MTDi distributions.
  , dist = "SDistribution"
  ),
  contains = c("mtdi_generator","VIRTUAL")
)
```

I was saving the generated `mtdi_distributions` as part of simulation objects, with one per simulation rep. Even moderately sized simulations (~ 10^2 – 10^3 reps) saved in the workspace were causing R restart times of many minutes. By profiling, I traced the overwhelming majority of this time to the invocation of `Lognormal$new` in this constructor:

```r
setMethod("initialize", "mtdi_lognormal",
  function(.Object, CV, median, ...) {
    .Object@dist <- Lognormal$new(meanlog = log(median),
                                  sdlog = sqrt(log(CV^2 + 1)))
    # Initialize 'bottom-up' to avoid a premature validity check:
    .Object <- callNextMethod(.Object, CV=CV, median=median, ...)
  })
```

This led me to try the following checks:
```r
> system.time(for(i in 1:1000) Lognormal$new(meanlog = 0, sdlog = 0.5))
   user  system elapsed 
 76.790   0.360  77.135 
> system.time(for(i in 1:1000) Normal$new(mean = 0, sd = 0.5))
   user  system elapsed 
 32.015   0.217  32.226 
```

In an email, Rafael has kindly mentioned `VectorDistribution` as one possible solution.
For now, however, I'll factor out the `@distr` slot of `mtdi_lognormal` and just hold on to the parameters and call the base functions. But as performance improves, I'll surely consider factoring distr6 back in!

**Comment by RaphaelS1** [comment]
Thanks David. I do find this use-case slightly strange because your wrapping R6 in S4, which is notoriously slow and clunky, that aside...there is a real problem in distr6 which I want to fix ASAP. In the future I'd also recommend using `microbenchmark` for timed tests.

In terms of bottleneck issues below is a reprex of a benchmark comparing an empty R6 class, the simplest learner in mlr3, construction of Normal and Lognormal distributions, construction of Normal and Lognormal parameter sets, and construction & setting of lognormal parameter set. Results clearly indicate that bottleneck lies in the parameter set and solving this would bring construction times down below that of the mlr3 learner.

``` r
library(distr6)
library(mlr3)
r = R6::R6Class("new")
microbenchmark::microbenchmark(
  r$new(),
  LearnerClassifFeatureless$new(),
  Normal$new(),
  Lognormal$new(),
  distr6:::getParameterSet.Normal(),
  distr6:::getParameterSet.Lognormal(mean = 1),
  distr6:::getParameterSet.Lognormal(mean = 1, var = 1)$setParameterValue(mean = 2)
)
#> Unit: microseconds
#>                                                                               expr
#>                                                                            r$new()
#>                                                    LearnerClassifFeatureless$new()
#>                                                                       Normal$new()
#>                                                                    Lognormal$new()
#>                                                  distr6:::getParameterSet.Normal()
#>                                       distr6:::getParameterSet.Lognormal(mean = 1)
#>  distr6:::getParameterSet.Lognormal(mean = 1, var = 1)$setParameterValue(mean = 2)
#>        min         lq       mean    median         uq       max neval   cld
#>     33.800    60.3065   834.8382    70.976     91.156  75622.12   100 a    
#>   3050.243  3518.1635  6511.6214  4279.809   5549.075 167752.03   100 a    
#>  29419.225 35104.7800 43745.5697 41436.095  47719.810 119293.17   100   c  
#>  69481.206 86348.8320 97266.3940 94616.133 108465.462 167235.13   100     e
#>  18201.166 21667.5700 30137.3080 26357.628  32899.551 173768.84   100  b   
#>  52782.906 64256.7585 74917.8338 73261.441  83525.861 162756.49   100    d 
#>  53195.053 70728.2035 79635.1855 76564.469  90403.260 117582.46   100    d
```

<sup>Created on 2020-07-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

<details>

<summary>Session info</summary>

``` r
devtools::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       macOS Catalina 10.15.3      
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_GB.UTF-8                 
#>  ctype    en_GB.UTF-8                 
#>  tz       Europe/London               
#>  date     2020-07-28                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package        * version    date       lib
#>  assertthat       0.2.1      2019-03-21 [1]
#>  backports        1.1.8      2020-06-17 [1]
#>  callr            3.4.3      2020-03-28 [1]
#>  checkmate        2.0.0      2020-02-06 [1]
#>  cli              2.0.2      2020-02-28 [1]
#>  codetools        0.2-16     2018-12-24 [1]
#>  crayon           1.3.4      2017-09-16 [1]
#>  data.table       1.13.0     2020-07-24 [1]
#>  desc             1.2.0      2018-05-01 [1]
#>  devtools         2.3.0      2020-04-10 [1]
#>  digest           0.6.25     2020-02-23 [1]
#>  distr6         * 1.4.2.9000 2020-07-28 [1]
#>  ellipsis         0.3.0      2019-09-20 [1]
#>  evaluate         0.14       2019-05-28 [1]
#>  fansi            0.4.1      2020-01-08 [1]
#>  fs               1.4.1      2020-04-04 [1]
#>  glue             1.4.1      2020-05-13 [1]
#>  highr            0.8        2019-03-20 [1]
#>  htmltools        0.4.0      2019-10-04 [1]
#>  knitr            1.28       2020-02-06 [1]
#>  lattice          0.20-41    2020-04-02 [1]
#>  lgr              0.3.4      2020-03-20 [1]
#>  magrittr         1.5        2014-11-22 [1]
#>  MASS             7.3-51.6   2020-04-26 [1]
#>  Matrix           1.2-18     2019-11-27 [1]
#>  memoise          1.1.0      2017-04-21 [1]
#>  microbenchmark   1.4-7      2019-09-24 [1]
#>  mlr3           * 0.4.0      2020-07-22 [1]
#>  mlr3misc         0.4.0      2020-07-17 [1]
#>  multcomp         1.4-13     2020-04-08 [1]
#>  mvtnorm          1.1-0      2020-02-24 [1]
#>  paradox          0.4.0      2020-07-21 [1]
#>  pkgbuild         1.0.7      2020-04-25 [1]
#>  pkgload          1.0.2      2018-10-29 [1]
#>  prettyunits      1.1.1      2020-01-24 [1]
#>  processx         3.4.2      2020-02-09 [1]
#>  ps               1.3.3      2020-05-08 [1]
#>  R6               2.4.1      2019-11-12 [1]
#>  R62S3            1.4.1      2020-03-09 [1]
#>  Rcpp             1.0.5      2020-07-06 [1]
#>  remotes          2.1.1      2020-02-15 [1]
#>  rlang            0.4.6      2020-05-02 [1]
#>  rmarkdown        2.2        2020-05-31 [1]
#>  rprojroot        1.3-2      2018-01-03 [1]
#>  sandwich         2.5-1      2019-04-06 [1]
#>  sessioninfo      1.1.1      2018-11-05 [1]
#>  set6             0.1.8      2020-07-27 [1]
#>  stringi          1.4.6      2020-02-17 [1]
#>  stringr          1.4.0      2019-02-10 [1]
#>  survival         3.1-12     2020-04-10 [1]
#>  testthat         2.3.2      2020-03-02 [1]
#>  TH.data          1.0-10     2019-01-21 [1]
#>  usethis          1.6.0      2020-04-09 [1]
#>  uuid             0.1-4      2020-02-26 [1]
#>  withr            2.2.0      2020-04-20 [1]
#>  xfun             0.13       2020-04-13 [1]
#>  yaml             2.2.1      2020-02-01 [1]
#>  zoo              1.8-7      2020-01-10 [1]
#>  source                                       
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  Github (alan-turing-institute/distr6@9cc7caf)
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.2)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#>  CRAN (R 4.0.0)                               
#> 
#> [1] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
```

</details>

**Comment by RaphaelS1** [comment]
~3-10x speed up.

``` r
library(distr6)
library(mlr3)
r = R6::R6Class("new")
microbenchmark::microbenchmark(
  r$new(),
  LearnerClassifFeatureless$new(),
  Normal$new(),
  Lognormal$new(),
  distr6:::getParameterSet.Normal(),
  distr6:::getParameterSet.Lognormal(mean = 1),
  distr6:::getParameterSet.Lognormal(mean = 1, var = 1)$setParameterValue(mean = 2)
)
#> Unit: microseconds
#>                                                                               expr
#>                                                                            r$new()
#>                                                    LearnerClassifFeatureless$new()
#>                                                                       Normal$new()
#>                                                                    Lognormal$new()
#>                                                  distr6:::getParameterSet.Normal()
#>                                       distr6:::getParameterSet.Lognormal(mean = 1)
#>  distr6:::getParameterSet.Lognormal(mean = 1, var = 1)$setParameterValue(mean = 2)
#>        min         lq      mean     median        uq       max neval  cld
#>     31.789    57.1645  1207.798    64.6005    71.987 113738.55   100 a   
#>   3083.991  3367.2975  5181.830  3763.2205  4819.421  83905.51   100 a   
#>  12764.134 14331.6965 18747.848 17217.6935 20993.096  37973.14   100  bc 
#>  23517.629 27873.7380 36120.273 32450.4630 38029.215 120395.37   100    d
#>   3653.503  3915.8235  5519.298  4318.4095  5426.181  50793.77   100 a   
#>   9358.475 10198.3670 14215.442 12770.9360 15660.808  47245.79   100  b  
#>  13110.659 14578.0095 21551.871 17091.8665 20224.137 284981.07   100   c
```

<sup>Created on 2020-07-28 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by RaphaelS1** [comment]
Results clean-up:
Times in microseconds

| Test | Original | Update 1 | Update 2 |
|------|---------|----------|----------|
| Empty R6 | 834 | 1207 | 77 |
| LearnerFeatureless | 6511 | 5181 | 4470 |
| Normal$new | 43745 | 18747 | 21599 |
| Lognormal$new | 97266 | 36120 | 31554 |
| Normal PS | 30137.3080 | 5519.298 | 5478 |
| Lognormal PS | 74917 | 14215 | 13279 |
| Lognormal Set PS | 79635 | 21552 | 21027 |

**Comment by dcnorris** [comment]
Many thanks, Rafael, and also for your tip about package [microbenchmark](https://CRAN.R-project.org/package=microbenchmark).

Here's what I did to factor out distr6 from the key classes without affecting the interface:

```r
setMethod("initialize", "mtdi_lognormal",
  function(.Object, CV, median, ...) {
    # .Object@dist <- Lognormal$new(meanlog = log(median),
    #                               sdlog = sqrt(log(CV^2 + 1)))
    meanlog <- log(median)
    sdlog <- sqrt(log(CV^2 + 1))
    .Object@dist <- list(
      quantile = function(p) qlnorm(p, meanlog = meanlog, sdlog = sdlog)
    , cdf = function(q) plnorm(q, meanlog = meanlog, sdlog = sdlog)
    , name = "Lognormal"
    )
    # Initialize 'bottom-up' to avoid a premature validity check:
    .Object <- callNextMethod(.Object, CV=CV, median=median, ...)
  })
```

Thus, I effectively 'spoof' the `$cdf/quantile/name` interface (which is all I use), leaving the other code intact. All subjectively apparent performance issues disappear. Objectively, results are:

```r
> microbenchmark(test_draw_samples(n = 100)) # new code
Unit: milliseconds
                       expr      min       lq     mean   median       uq      max neval
 test_draw_samples(n = 100) 43.55572 44.61792 50.57526 45.33719 52.10775 113.7552   100
```

This shows about 0.5ms (500µs) timing per sample.
This refactoring does, to be sure, lose the expressiveness of your class hierarchy as I no longer have a virtual superclass with @dist slot of type "SDistribution".

**Comment by RaphaelS1** [comment]
Closed in #255
